File: cdm/java/mil/tatrc/physiology/datamodel/properties/SEScalarOsmolarity.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.properties;

import com.kitware.physiology.cdm.Properties.ScalarOsmolarityData;

import mil.tatrc.physiology.datamodel.properties.CommonUnits.OsmolarityUnit;

/**
 * @author abray
 * Osmolarity Scalar Class, will enforce that units are proper to Osmolarity
 */

public class SEScalarOsmolarity extends SEScalar
{
  public SEScalarOsmolarity()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarOsmolarity(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarOsmolarity(double value, OsmolarityUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarOsmolarityData src, SEScalarOsmolarity dst)
  {
    SEScalar.load(src.getScalarOsmolarity(),dst);
  }
  public static ScalarOsmolarityData unload(SEScalarOsmolarity src)
  {
    if(!src.isValid())
      return null;
    ScalarOsmolarityData.Builder dst = ScalarOsmolarityData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarOsmolarity src, ScalarOsmolarityData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarOsmolarityBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, OsmolarityUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(OsmolarityUnit unit)
  {
    return this.getValue(unit.toString());
  }

  public boolean validUnit(String unit)
  {
    if(OsmolarityUnit.validUnit(unit))
      return true;
    return false;
  }
}

File: cdm/cpp/properties/SEScalarOsmolarity.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarOsmolarity.h"

const OsmolarityUnit OsmolarityUnit::Osm_Per_L("Osm/L");
const OsmolarityUnit OsmolarityUnit::mOsm_Per_L("mOsm/L");

bool OsmolarityUnit::IsValidUnit(const std::string& unit)
{
  if (Osm_Per_L.GetString().compare(unit) == 0)
    return true;
  if (mOsm_Per_L.GetString().compare(unit) == 0)
    return true;
  return false;
}

const OsmolarityUnit& OsmolarityUnit::GetCompoundUnit(const std::string& unit)
{
  if (Osm_Per_L.GetString().compare(unit) == 0)
    return Osm_Per_L;
  if (mOsm_Per_L.GetString().compare(unit) == 0)
    return mOsm_Per_L;
  std::stringstream err;
  err << unit << " is not a valid Osmolarity unit";
  throw CommonDataModelException(err.str());
}

void SEScalarOsmolarity::Load(const cdm::ScalarOsmolarityData& src, SEScalarOsmolarity& dst)
{
  SEScalarOsmolarity::Serialize(src, dst);
}
void SEScalarOsmolarity::Serialize(const cdm::ScalarOsmolarityData& src, SEScalarOsmolarity& dst)
{
  SEScalarQuantity<OsmolarityUnit>::Serialize(src.scalarosmolarity(), dst);
}

cdm::ScalarOsmolarityData* SEScalarOsmolarity::Unload(const SEScalarOsmolarity& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarOsmolarityData* dst = new cdm::ScalarOsmolarityData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarOsmolarity::Serialize(const SEScalarOsmolarity& src, cdm::ScalarOsmolarityData& dst)
{
  SEScalarQuantity<OsmolarityUnit>::Serialize(src, *dst.mutable_scalarosmolarity());
}